Migrate to register(storeDescriptor) for shipping zone store data store

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/55390

As part of the React 18 upgrade and WordPress 6.6 compatibility changes, this PR updates the **experimental shipping zones** store registration pattern in `@woocommerce/data` to use the modern WordPress data store API. 

Key changes include:

1. Migrated from deprecated `registerStore` to `createReduxStore` and `register` pattern
2. Updated store exports to match the modern pattern used across other stores:
    - Added `export { store as experimentalShippingZonesStore }`
    - Maintained `EXPERIMENTAL_SHIPPING_ZONES_STORE_NAME` export for backward compatibility
3. Removed deprecated type definitions:
    - Removed custom @wordpress/data module declaration
    - Cleaned up unused type exports and imports
4. Update experimental shipping zones store import to use `experimentalShippingZonesStore` from `@woocommerce/data`

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR doesn't change any functionality, so it's not necessary to test the changes.

For devs:

1. Review the @woocommerce/data changes for the shipping zones store migration
2. Delete the `noCheck: true` line in `./packages/js/data/tsconfig.json` and running `npx tsc` inside ./packages/js/data
3. Confirm that there are no TS errors in `./src/shipping-zones/**`
4. `window.wp.data.select( wc.data.experimentalShippingZonesStore )` and `window.wp.data.dispatch( wc.data.experimentalShippingZonesStore )` should return selectors and dispatchers

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Migrate shipping zones data store to use register(storeDescriptor)

Update shipping zones data store to use the new createReduxStore and register pattern, similar to other data stores. This includes:
- Exporting the store from the shipping zones data store
- Updating the index to expose the experimental shipping zones store
- Updating the README with the new store usage pattern

chihsuan committed:
Add changelog

chihsuan committed:
Fix lint

chihsuan committed:
Remove unused import

chihsuan committed:
Fix types

chihsuan committed:
Simplify createCrudDataStore return value in shipping zones and crud stores

ilyasfoo commented on packages/js/data/src/shipping-zones/README.md:
```diff
@@ -13,20 +13,19 @@ Example:
 
```
I think it'll be good update here as well: https://github.com/woocommerce/woocommerce/blob/3b92b879c7ede470b2f531189389424340f3d417/packages/js/data/src/shipping-zones/README.md?plain=1#L10

chihsuan replied:
Thanks, updated it in a0dfef5f1d74e5a9ebc39ae7345cbd93748256b9! 🙏

ilyasfoo reviewed (approved):
Tests well, but with minor comment on documentation. Pre-approving!

chihsuan committed:
Update shipping zones README with new store export constant

chihsuan committed:
Merge branch 'trunk' into fix/migrate-shipping-zone-store

chihsuan committed:
Migrate to register(storeDescriptor) for  tax classes store (#55471)

* Export experimental tax classes store

Add export for the experimental tax classes store in the data package index, making it accessible for use in other parts of the application.

* Add changelog

* Update types

adrianduffell reviewed (approved):
LGTM and testing well! 🚀
